x/mongo/driver/auth/scram_test.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (C) MongoDB, Inc. 2017-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package auth
+
+import (
+	"context"
+	"testing"
+
+	"go.mongodb.org/mongo-driver/internal/testutil/assert"
+	"go.mongodb.org/mongo-driver/x/bsonx/bsoncore"
+	"go.mongodb.org/mongo-driver/x/mongo/driver/description"
+	"go.mongodb.org/mongo-driver/x/mongo/driver/drivertest"
+)
+
+const (
+	scramSha1Nonce   = "fyko+d2lbbFgONRv9qkxdawL"
+	scramSha256Nonce = "rOprNGfwEbeRWgbNEkqO"
+)
+
+var (
+	scramSha1ShortPayloads = [][]byte{
+		[]byte("r=fyko+d2lbbFgONRv9qkxdawLHo+Vgk7qvUOKUwuWLIWg4l/9SraGMHEE,s=rQ9ZY3MntBeuP3E1TDVC4w==,i=10000"),
+		[]byte("v=UMWeI25JD1yNYZRMpZ4VHvhZ9e0="),
+	}
+	scramSha256ShortPayloads = [][]byte{
+		[]byte("r=rOprNGfwEbeRWgbNEkqO%hvYDpWUa2RaTCAfuxFIlj)hNlF$k0,s=W22ZaJ0SNY7soEsUEjb6gQ==,i=4096"),
+		[]byte("v=6rriTRBi23WpRR/wtup+mMhUZUn/dB5nLTJRsjl95G4="),
+	}
+	scramSha1LongPayloads   = append(scramSha1ShortPayloads, []byte{})
+	scramSha256LongPayloads = append(scramSha256ShortPayloads, []byte{})
+)
+
+func TestSCRAM(t *testing.T) {
+	t.Run("conversation", func(t *testing.T) {
+		testCases := []struct {
+			name                  string
+			createAuthenticatorFn func(*Cred) (Authenticator, error)
+			payloads              [][]byte
+			nonce                 string
+		}{
+			{"scram-sha-1 short conversation", newScramSHA1Authenticator, scramSha1ShortPayloads, scramSha1Nonce},
+			{"scram-sha-256 short conversation", newScramSHA256Authenticator, scramSha256ShortPayloads, scramSha256Nonce},
+			{"scram-sha-1 long conversation", newScramSHA1Authenticator, scramSha1LongPayloads, scramSha1Nonce},
+			{"scram-sha-256 long conversation", newScramSHA256Authenticator, scramSha256LongPayloads, scramSha256Nonce},
+		}
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				authenticator, err := tc.createAuthenticatorFn(&Cred{
+					Username: "user",
+					Password: "pencil",
+					Source:   "admin",
+				})
+				assert.Nil(t, err, "newScramSHA1Authenticator error: %v", err)
```

Comment:
Isn't this sometimes a newScramSHA256Authenticator error?

Reply:
Whoops. This was just a SCRAM-SHA-1 test before I made it table driven. Changed the message to be more generic.